haskell/protobuf.bzl

```diff
@@ -0,0 +1,246 @@
+"""Support for protocol buffers"""
+
+load(":providers.bzl", "HaskellBuildInfo", "HaskellProtobufInfo")
+load("@bazel_skylib//:lib.bzl", "dicts", "paths")
+load(":tools.bzl", "protobuf_tools")
+load(":path_utils.bzl", "target_unique_name")
+
+def _capitalize_first_letter(c):
+  """Capitalize the first letter of the input. Unlike the built-in
+  `capitalize()` method, doesn't lower-case the other characters. This helps
+  mimic the behavior of `proto-lens-protoc`, which turns `Foo/Bar/BAZ.proto`
+  into `Foo/Bar/BAZ.hs` (rather than `Foo/Bar/Baz.hs`).
+
+  Args:
+    c: A non-empty string word.
+
+  Returns:
+    The input with the first letter upper-cased.
+  """
+  return c[0].capitalize() + c[1:]
+
+def _camel_case(comp):
+  """Camel-case the input string, preserving any existing capital letters.
+  """
+  # Split on both "-" and "_", matching the behavior of proto-lens-protoc.
+  return "".join([_capitalize_first_letter(c2)
+                  for c1 in comp.split("_") for c2 in c1.split("-")])
+
+def _proto_lens_output_file(path):
+  """The output file from `proto-lens-protoc` when run on the given `path`.
+  """
+
+  path = path[:-len(".proto")]
+  result = "/".join([_camel_case(p) for p in path.split("/")]) + ".hs"
+
+  return "Proto/" + result
+
+def _haskell_proto_aspect_impl(target, ctx):
+
+  args = ctx.actions.args()
+
+  src_prefix = paths.join(
+    ctx.label.workspace_root,
+    ctx.label.package,
+  )
+
+  args.add(["--plugin=protoc-gen-haskell=" +
+            protobuf_tools(ctx).protoc_gen_haskell.path,
+  ])
+
+  hs_files = []
+  inputs = []
+
+  args.add(["-I{0}={1}".format(s.short_path, s.path)
+            for s in target.proto.transitive_sources])
+
+  inputs.extend(target.proto.transitive_sources.to_list())
```

Comment:
This seems like it will generate a single haskell_library out of all the transitive dependencies.  Compare to `cabal2bazel` which only builds the direct_sources: https://github.com/google/cabal2bazel/blob/c208bf179c8e820b0733a4472a5825c4f86e8e60/bzl/def.bzl#L1093
which generates a `haskell_library` out of only the immediate sources, but then makes it depend on the corresponding `proto_library`/aspect deps.

(See below comments on the `BUILD` file.)